daml-lf/transaction-scalacheck/src/main/scala/com/digitalasset/daml/lf/value/TypedValueGenerators.scala

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf
+package value
+
+import scala.language.higherKinds
+import data.{Decimal, FrontStack, Ref, SortedLookupList, Time}
+import data.ImmArray.ImmArraySeq
+import iface.{PrimType => PT, Type, TypePrim}
+
+import scalaz.Id.Id
+import scalaz.syntax.traverse._
+import scalaz.std.option._
+import org.scalacheck.{Arbitrary, Gen, Shrink}
+import Arbitrary.arbitrary
+
+/** [[ValueGenerators]] produce untyped values; for example, if you use the list gen,
+  * you get a heterogeneous list.  The generation target here, on the other hand, is
+  * ''a pair of a type and a generator of values of that type''.  For example, you might
+  * generate the type `[Map Decimal]`, which would be accompanied by a `Gen` that produced
+  * [[Value]]s each guaranteed to be a list of maps, whose values are all guaranteed to
+  * be Decimals.
+  *
+  * As a user, you will probably be most interested in one of the generators derived
+  * from `genAddend`; if you need a generator in this theme not already supported by
+  * one such generator, you can probably derive a new one from `genAddend` yourself.
+  */
+object TypedValueGenerators {
+  sealed abstract class ValueAddend {
+    type Inj[Cid]
+    def t: Type
+    def inj[Cid]: Inj[Cid] => Value[Cid]
+    def prj[Cid]: Value[Cid] => Option[Inj[Cid]]
+    def injgen[Cid](cid: Gen[Cid]): Gen[Inj[Cid]]
+    implicit def injshrink[Cid: Shrink]: Shrink[Inj[Cid]]
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  object ValueAddend extends PrimInstances[Lambda[a => ValueAddend { type Inj[_] = a }]] {
+    type Aux[Inj0[_]] = ValueAddend {
+      type Inj[Cid] = Inj0[Cid]
+    }
+    type NoCid[Inj0] = ValueAddend {
+      type Inj[_] = Inj0
+    }
+
+    def noCid[Inj0](pt: PT, inj0: Inj0 => Value[Nothing])(prj0: Value[Any] PartialFunction Inj0)(
+        implicit arb: Arbitrary[Inj0],
+        shr: Shrink[Inj0]): NoCid[Inj0] = new ValueAddend {
+      type Inj[Cid] = Inj0
+      override val t = TypePrim(pt, ImmArraySeq.empty)
+      override def inj[Cid] = inj0
+      override def prj[Cid] = prj0.lift
+      override def injgen[Cid](cid: Gen[Cid]) = arb.arbitrary
+      override def injshrink[Cid: Shrink] = shr
+    }
+
+    import Value._, ValueGenerators.Implicits._
+    val text = noCid(PT.Text, ValueText) { case ValueText(t) => t }
+    val int64 = noCid(PT.Int64, ValueInt64) { case ValueInt64(i) => i }
+    val decimal = noCid(PT.Decimal, ValueDecimal) { case ValueDecimal(d) => d }
+    val unit = noCid(PT.Unit, (_: Unit) => ValueUnit) { case ValueUnit => () }
+    val date = noCid(PT.Date, ValueDate) { case ValueDate(d) => d }
+    val timestamp = noCid(PT.Timestamp, ValueTimestamp) { case ValueTimestamp(t) => t }
+    val bool = noCid(PT.Bool, ValueBool) { case ValueBool(b) => b }
+    val party = noCid(PT.Party, ValueParty) { case ValueParty(p) => p }
+
```

Comment:
what about `PT.List` and `PT.Optional`

Reply:
and also `PT.Map`

Reply:
Not possible to do with `noCid`, scroll down for support for those three, and also contractId.